---
title: "Meta-analysis"
output: html_document
date: "2025-02-26"
---
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

Packages used: 

```{r}
library(Seurat)
library(Matrix)
library(biomaRt)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(ggplot2)
library(dplyr)
library(stringr)
library(glmGamPoi)
library(readxl)
library(EnhancedVolcano)
library(speckle)
library(presto)
library(MAST)
library(openxlsx)
library(patchwork)
library(tibble)
```

```{r}
load(file = "~/R/Data/Seurats_Acute.RData")
load(file = "~/R/Data/Seurats_Chronic.RData")
rm(seurat_list, seurat_list_norm)
```

```{r}
Idents(seurat) <- "time"
Idents(seurat2) <- "time"
Idents(seurat3) <- "time"
Idents(seurat4) <- "time"
Idents(seurat5) <- "time"
```

```{r}
seurat_list <- list(seurat, seurat2, seurat3, seurat4, seurat5)
```

UMAP before CCA integration, only merge 

```{r}
combined_seurat <- merge(
  x = seurat_list[[1]],
  y = list(seurat_list[[2]], seurat_list[[3]], seurat_list[[4]], seurat_list[[5]]), project = "MergedProject")

# Pre-process RNA assay to see clustering before integration
DefaultAssay(combined_seurat) <- "RNA"

combined_seurat <- JoinLayers(combined_seurat)
combined_seurat <- NormalizeData(combined_seurat)
combined_seurat <- FindVariableFeatures(combined_seurat)
gc()
combined_seurat <- ScaleData(combined_seurat)
gc()
combined_seurat <- RunPCA(combined_seurat, dims = 1:30)
gc()
combined_seurat <- RunUMAP(combined_seurat, dims = 1:30)

umap_meta <- DimPlot(combined_seurat, group.by = "dataset") + theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 7), legend.title = element_text(size = 20), legend.text = element_text(size = 16)) + labs(title = "Pre-integration UMAP clustering by dataset", x = "UMAP 1", y = "UMAP 2")

ggsave("~/R/Figures2/UMAP_BeforeIntegration_MetaAnalysis2.png", umap_meta, dpi = 1080, height = 8, width = 9)

rm(seurat2_norm, seurat_norm, seurat3_norm, seurat4_norm, seurat5_norm)

```

Integration

```{r}

for (i in seq_along(seurat_list)) {
  seurat_list[[i]] <- NormalizeData(seurat_list[[i]], normalization.method = "LogNormalize")
  seurat_list[[i]] <- FindVariableFeatures(seurat_list[[i]], selection.method = "vst", nfeatures = 3000)
}

features <- SelectIntegrationFeatures(
  object.list = seurat_list,
  nfeatures = 3000)

for (i in seq_along(seurat_list)) {
  seurat_list[[i]] <- ScaleData(seurat_list[[i]], features = features)
  seurat_list[[i]] <- RunPCA(seurat_list[[i]], features = features)
  seurat_list[[i]] <- RunUMAP(seurat_list[[i]], features = features)
}

gc()
anchors_integration <- FindIntegrationAnchors(object.list = seurat_list, dims = 1:30, verbose = TRUE, reduction = "cca", anchor.features = features, scale = F, normalization.method = "LogNormalize", k.anchor = 10)

rm(seurat_list)

save.image(file = "~/R/Downloads/AllData_Anchors.RData")
```

Integration 

```{r}
load(file = "~/R/Downloads/AllData_Anchors.RData")
gc()
seurat_integrated <- IntegrateData(anchorset = anchors_integration, dims = 1:30, verbose = TRUE, normalization.method = "LogNormalize") 

rm(seurat_list, anchors, anchors_integration)

save.image(file = "~/R/Downloads/AllData_CCA.RData")
```

Load integrated dataset

```{r}
load(file = "~/R/Downloads/AllData_CCA.RData")
gc()
```

```{r}
# Check Integration
gc()
seurat_integrated <- ScaleData(seurat_integrated)
gc()
seurat_integrated <- RunPCA(seurat_integrated, verbose = T, dims = 1:30) 
gc()
seurat_integrated <- RunUMAP(seurat_integrated, dims = 1:30)


datasets <- DimPlot(seurat_integrated, group.by = "dataset", label = F) + labs(title = "Post-integration UMAP clustering by dataset", x = "UMAP 1", y = "UMAP 2") + theme(
  legend.title = element_text(size = 16),
  legend.text = element_text(size = 16)
)

datasets

ggsave("~/R/Figures2/Datasets_MetaAnalysis.png", datasets, dpi = 1080, height = 8, width = 9)
```

Cell type annotation - Transfer Learning

```{r}
seurat_integrated@meta.data <- seurat_integrated@meta.data %>%
  dplyr::select(-starts_with("pred"))

# Subset the integrated object into reference (Skinnider_2024) and query (all others)
reference <- subset(seurat_integrated, subset = dataset == "Skinnider_2024")
query <- subset(seurat_integrated, subset = dataset != "Skinnider_2024")

# Find anchors and transfer labels from reference to query
anchors <- FindTransferAnchors(reference = reference, query = query, dims = 1:30)
predictions <- TransferData(anchorset = anchors, refdata = reference$celltype, dims = 1:30)
query <- AddMetaData(query, metadata = predictions)
query$predicted_celltype <- predictions$predicted.id

# Filter by prediction score
nrow(query@meta.data)
table(query$prediction.score.max > 0.9)
query <- subset(query, subset = prediction.score.max > 0.9)

# Create a new metadata column in the full object
seurat_integrated$predicted_celltype <- NA

# Assign original cell types for Skinnider_2024 cells
seurat_integrated$predicted_celltype[Cells(reference)] <- reference$celltype

# Assign predicted cell types for the rest
seurat_integrated$predicted_celltype[Cells(query)] <- query$predicted_celltype

# Delete NAs in celltype
rm(query, reference)
cells_to_keep <- rownames(seurat_integrated@meta.data)[!is.na(seurat_integrated@meta.data$predicted_celltype)]
seurat_integrated <- subset(seurat_integrated, cells = cells_to_keep)
 # No celltype label - because no prediction.score > 0.9
```


```{r}
# Accuracy

mapping <- list(
  "Astrocyte" = "Astrocytes",
  "Astrocytes" = "Astrocytes",
  "B cells" = c("B cells","Peripheral immune cells"),
  "Dendritic" = c("Peripheral immune cells", "B cells", "NK, T cells", "Myeloid, dividing"),
  "Div-Myeloid" = "Myeloid, dividing",
  "Dorsal" = "Dorsal",
  "Endothelial" = "Vascular endothelial cells",
  "Endothelial cells" = "Vascular endothelial cells",
  "Ependymal" = "Ependymal cells",
  "Ependymal cells" = "Ependymal cells",
  "Fibroblast" = "Vascular cells",
  "Leptomeninges" = "Vascular cells",
  "Lymphocyte" = c("Peripheral immune cells", "B cells", "NK, T cells"),
  "Macrophage" = c("Peripheral immune cells", "Myeloid, dividing"),
  "Mature oligodendrocytes (MOL)" = "Mature oligodendrocytes (MOL)",
  "Microglia" = "Microglia",
  "Microglia/Hematopoietic" = c("Microglia","Peripheral immune cells", "Myeloid, dividing"),
  "Monocyte" = c("Peripheral immune cells", "Myeloid, dividing"),
  "Myelin forming oligodendrocytes (MFOL)" = "Myelin forming oligodendrocytes (MFOL)",
  "Myeloid, dividing" = "Myeloid, dividing",
  "Neuron" = c("Dorsal", "Ventral"),
  "Neurons" = c("Dorsal", "Ventral"),
  "Neutrophil" = c("Peripheral immune cells", "Myeloid, dividing"),
  "Newly formed oligodendrocytes (NFOL)" = "Newly formed oligodendrocytes (NFOL)",
  "NK, T cells" = c("Peripheral immune cells", "NK, T cells"),
  "Oligodendrocyte" = c("Newly formed oligodendrocytes (NFOL)", "Mature oligodendrocytes (MOL)", "Myelin forming oligodendrocytes (MFOL)"),
  "Oligodendrocytes" = c("Newly formed oligodendrocytes (NFOL)", "Mature oligodendrocytes (MOL)", "Myelin forming oligodendrocytes (MFOL)"),
  "Oligodendrocyte Progenitors/Precursors" = "Oligodendrocytes precursor cells (OPC)",
  "Oligodendrocytes precursor cells (OPC)" = "Oligodendrocytes precursor cells (OPC)",
  "OPC" = "Oligodendrocytes precursor cells (OPC)",
  "OPCs" = "Oligodendrocytes precursor cells (OPC)",
  "Pericyte" = "Vascular cells",
  "Pericytes" = "Vascular cells",
  "Peripheral immune cells" = c("Peripheral immune cells","Myeloid, dividing"),
  "Vascular cells" = "Vascular cells",
  "Vascular endothelial cells" = "Vascular endothelial cells",
  "Ventral" = "Ventral"
)

# Matches
check_manual_match <- function(original, predicted) {
  original <- as.character(original)
  predicted <- as.character(predicted)
  
  if (!original %in% names(mapping)) {
    return(FALSE)
  }
  
  valid_matches <- mapping[[original]]
  if (length(valid_matches) == 0) {
    return(FALSE)
  }
  
  return(predicted %in% valid_matches)
}

metadata <- seurat_integrated@meta.data
metadata$is_match <- apply(metadata, 1, function(row) {
  check_manual_match(row["celltype"], row["predicted_celltype"])
})

# Accuray of the annotation
accuracy <- mean(metadata$is_match, na.rm = TRUE)
accuracy

# Filter cells bad annotated
discrepancies <- metadata[!metadata$is_match, c("celltype", "predicted_celltype")]
cells_to_remove <- rownames(discrepancies)
gc()
seurat_integrated <- seurat_integrated[, !colnames(seurat_integrated) %in% cells_to_remove]

# UMAP Plot for cell types
seurat_integrated <- RunUMAP(seurat_integrated, dims = 1:30)

celltype <- DimPlot(seurat_integrated, group.by = "predicted_celltype", label = T, label.size = 5.5, repel = T) + labs(x = "UMAP 1", y = "UMAP 2", caption = "Cells with prediction score > 0.9
                                                                                          Transfer learning accuracy = 97.22%") +
ggtitle("                                      Cell type annotation after transfer learning approach") +
  theme(plot.title = element_text(hjust = 0.5, size = 18),
  legend.text = element_text(size = 14)
)
celltype

ggsave("~/R/Figures2/Cells_MetaAnalysis.png", celltype, dpi = 1080, height = 9, width = 13)

```

* METADATA *

```{r}

metadata <- seurat_integrated@meta.data
metadata <- metadata %>% dplyr::select(matches("nCount_RNA|nFeature_RNA|time|celltype|technology|dataset|predicted_celltype"))

metadata$time[metadata$time == "Sub-Acute"] <- "21 dpi"
metadata$time[metadata$time == "Sub-Chronic"] <- "4-6 wpi"

metadata$celltype <- metadata$predicted_celltype

# Return check metadata into seurat object
seurat_integrated@meta.data <- metadata

Idents(seurat_integrated) <- "time"

# Descriptive
table(seurat_integrated$celltype)
table(seurat_integrated$time)
table(seurat_integrated$dataset)

# We have a few number of cells within some cell types like B cells
# We lose specificity but we win reliability
# The vast majority of discrepancies are between near clusters: Highlight ! peripherial immune cells - microglia, oligodendrocytes - OPCs 

seurat_integrated@meta.data <- seurat_integrated@meta.data %>%
  mutate(celltype_grouped = case_when(
    celltype %in% c("Mature oligodendrocytes (MOL)", "Myelin forming oligodendrocytes (MFOL)") ~ "Oligodendrocytes",
    celltype == "Newly formed oligodendrocytes (NFOL)" ~ "Newly formed oligodendrocytes",
    celltype == "Oligodendrocytes precursor cells (OPC)" ~ "OPCs",
    celltype == "Astrocytes" ~ "Astrocytes",
    celltype == "Ependymal cells" ~ "Ependymal cells",
    celltype == "Microglia" ~ "Microglia",
    celltype %in% c("B cells", "NK, T cells", "Peripheral immune cells") ~ "Immune cells",
    celltype == "Vascular endothelial cells" ~ "Endothelial cells",
    celltype == "Vascular cells" ~ "Vascular cells",
    celltype == "Myeloid, dividing" ~ "Myeloid",
    celltype %in% c("Ventral", "Dorsal") ~ "Neurons",
    TRUE ~ "Other"
  ))

seurat_integrated@meta.data$celltype <- seurat_integrated@meta.data$celltype_grouped


```

UMAP Cells Grouped

```{r}

celltype <- DimPlot(seurat_integrated, group.by = "celltype", label = T, label.size = 5.5, repel = T) + labs(title = "                              Final cell type annotation ", x = "UMAP 1", y = "UMAP 2") + theme(plot.title = element_text(hjust = 0.5, size = 18), legend.text = element_text(size = 14)
)
celltype

ggsave("~/R/Figures2/CellsGrouped_MetaAnalysis.png", celltype, dpi = 1080, height = 9, width = 13)

```

Cell Cycle Score

```{r}

cell_cycle_markers <- read.csv("~/R/Data/Mus_musculus.csv")
s_genes <- cell_cycle_markers %>% dplyr::filter(phase == "S") %>% pull("geneID")
g2m_genes <- cell_cycle_markers %>% dplyr::filter(phase == "G2/M") %>% pull("geneID")
s_genes <- AnnotationDbi::select(org.Mm.eg.db,
                                   keys = s_genes,
                                   keytype = "ENSEMBL",
                                   columns = c("SYMBOL"))
g2m_genes <- AnnotationDbi::select(org.Mm.eg.db,
                                   keys = g2m_genes,
                                   keytype = "ENSEMBL",
                                   columns = c("SYMBOL"))

seurat_integrated <- CellCycleScoring(seurat_integrated, g2m.features = g2m_genes$SYMBOL, s.features = s_genes$SYMBOL)

seurat_integrated@meta.data %>%
  group_by(Phase) %>%
  summarise(
    S_min      = min(S.Score, na.rm = TRUE),
    S_median   = median(S.Score, na.rm = TRUE),
    S_max      = max(S.Score, na.rm = TRUE),
    G2M_min    = min(G2M.Score, na.rm = TRUE),
    G2M_median = median(G2M.Score, na.rm = TRUE),
    G2M_max    = max(G2M.Score, na.rm = TRUE)
  )

seurat_integrated <- subset(seurat_integrated, subset = Phase != "Undecided")

cc_score <- DimPlot(seurat_integrated,
        reduction = "pca",
        group.by= "Phase", split.by = "Phase") + labs(title = "Cell Cycle Score Assessment", x = "PCA 1", y = "PCA 2") + theme(
  legend.title = element_text(size = 16),
  legend.text = element_text(size = 16)
)
cc_score

cc_score2 <- DimPlot(seurat_integrated,
        reduction = "pca",
        group.by= "Phase") + labs(title = "Cell Cycle Score Assessment", x = "PCA 1", y = "PCA 2") + theme(
  legend.title = element_text(size = 16),
  legend.text = element_text(size = 16)
)
cc_score2

library(patchwork)

cc <- cc_score2 + cc_score
cc

ggsave("~/R/Figures2/CellCycleScore2.png", cc, dpi = 1080, height = 8, width = 16)

table(seurat_integrated@meta.data$Phase)

# It is normal that dividing oligodendrocytes have more G2M genes expressed, the other cell types are well mixed
```

```{r}
summary <- as.data.frame.array(table(seurat_integrated$celltype, seurat_integrated$dataset))
write.xlsx(summary, "~/R/Downloads/Descriptive.xlsx")

```

*Differential expression analysis* -> RNA assay: no integrated

```{r}
# Pre-process RNA assay to see clustering before integration
DefaultAssay(seurat_integrated) <- "RNA"

seurat_integrated <- JoinLayers(seurat_integrated)
seurat_integrated <- NormalizeData(seurat_integrated)

```

Filter ECM genes

```{r}
proteins.ME <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")
genes_to_keep <- proteins.ME$Gene
rownames_seurat <- rownames(seurat_integrated)
rows_to_keep <- which(rownames_seurat %in% genes_to_keep)
seurat_integrated_ME <- seurat_integrated[rows_to_keep, ]

paste("ECM Genes present in the dataset:", nrow(seurat_integrated_ME))

saveRDS(seurat_integrated_ME, "~/R/Downloads/Seurat_Integrated_ME.rds")

```

DE analysis

```{r}

Idents(seurat_integrated_ME) <- "time"

DEG <- FindMarkers(seurat_integrated_ME,  ident.1 = "7dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST", layer = "data", latent.vars = c("dataset", "nCount_RNA"))

DEG$Gene <- rownames(DEG)
DEG$p_val_adj <- ifelse(DEG$p_val_adj == 0, 1e-305, DEG$p_val_adj)
DEG$Time <- "7 dpi"

write.xlsx(DEG, "~/R/Downloads/DEG_Acute.xlsx")

########################################################################################################

DEG2 <- FindMarkers(seurat_integrated_ME,  ident.1 = "1dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST", layer = "data", latent.vars = c("dataset", "nCount_RNA"))

DEG2$Gene <- rownames(DEG2)
DEG2$p_val_adj <- ifelse(DEG2$p_val_adj == 0, 1e-305, DEG2$p_val_adj)
DEG2$Time <- "1 dpi"

write.xlsx(DEG2, "~/R/Downloads/DEG_FirstDay.xlsx")

##############################################################################################################

DEG3 <- FindMarkers(seurat_integrated_ME,  ident.1 = "4-6 wpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST", layer = "data", latent.vars = c("dataset", "nCount_RNA"))

DEG3$Gene <- rownames(DEG3)
DEG3$p_val_adj <- ifelse(DEG3$p_val_adj == 0, 1e-305, DEG3$p_val_adj)
DEG3$Time <- "4-6 wpi"

write.xlsx(DEG3, "~/R/Downloads/DEG_Sub-Chronic.xlsx")

##############################################################################################################

DEG4 <- FindMarkers(seurat_integrated_ME,  ident.1 = "21 dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST", layer = "data", latent.vars = c("dataset", "nCount_RNA"))

DEG4$Gene <- rownames(DEG4)
DEG4$p_val_adj <- ifelse(DEG4$p_val_adj == 0, 1e-305, DEG4$p_val_adj)
DEG4$Time <- "21 dpi"

write.xlsx(DEG4, "~/R/Downloads/DEG_21dpi.xlsx")

##############################################################################################################

```

```{r}

DEG_21dpi <- read_excel("~/R/Downloads/DEG_21dpi.xlsx")
DEG_1mpi <- read_excel("~/R/Downloads/DEG_Sub-Chronic.xlsx")
DEG_1dpi <- read_excel("~/R/Downloads/DEG_FirstDay.xlsx")
DEG_7dpi <- read_excel("~/R/Downloads/DEG_Acute.xlsx")

# DEGs <- bind_rows(DEG, DEG2, DEG3, DEG4)
DEGs <- bind_rows(DEG_1dpi, DEG_7dpi, DEG_21dpi, DEG_1mpi)

DEGs <- DEGs %>%
  filter(p_val_adj < 0.05)

peaks_abs <- DEGs %>%
  mutate(abs_log2FC = abs(avg_log2FC)) %>%
  group_by(Gene) %>%
  filter(abs_log2FC == max(abs_log2FC)) %>%
  slice(1) %>%  
  ungroup()

peaks_filtered <- peaks_abs %>%
  filter(abs_log2FC > 0.585)

counts_abs <- peaks_filtered %>%
  group_by(Time) %>%
  summarise(n_genes = n(), .groups = "drop")

time_order <- c("1 dpi", "7 dpi", "21 dpi", "4-6 wpi")

counts_abs$Time <- factor(counts_abs$Time, levels = time_order)

histo <- ggplot(counts_abs, aes(x = Time, y = n_genes)) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_text(aes(label = n_genes), vjust = -0.5, size = 5) +
  labs(
    x = "Time Point",
    y = "Number of Genes with Expression Peak",
    fill = "Direction"
  ) +
  theme_minimal() +
  theme(
    axis.title.x = element_text(size = 18),
    axis.title.y = element_text(size = 18),
    axis.text.x = element_text(size = 16),
    axis.text.y = element_text(size = 16)
  )

ggsave("~/R/Figures2/Histogram_PeakExpression2.png", histo, width = 10, height = 10,  dpi = 1080)

``` 



