---
title: "Other analysis"
output: html_document
date: "2025-04-03"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(readxl)
library(ggplot2)
library(ggvenn)
library(dplyr)
```

## Once I have the DEG list, I can make different analysis to correlate the results with proteomic data.

*ACUTE CONDITION (7 dpi)*

```{r}

# VENN DIAGRAMS 
transcriptomic <- read_excel("~/R/Downloads/DEG_Acute.xlsx")
proteomic <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")

transcriptomic_up <- transcriptomic %>% filter(transcriptomic$avg_log2FC > 0.5 & p_val_adj < 0.05)
transcriptomic_up <- transcriptomic_up$Gene
transcriptomic_down <- transcriptomic %>% filter(transcriptomic$avg_log2FC < -0.5 & p_val_adj < 0.05)
transcriptomic_down <- transcriptomic_down$Gene

proteomic_up <- proteomic %>% filter(expr_Adjpval_Acute_Ctrl == "up" & as.numeric(logFC_Acute_Ctrl) > 0.5)
proteomic_up <- proteomic_up$Gene
proteomic_down <- proteomic %>% filter(expr_Adjpval_Acute_Ctrl == "down" & as.numeric(logFC_Acute_Ctrl) < -0.5)
proteomic_down <- proteomic_down$Gene

x = list("Genes upregulated" = transcriptomic_up, "Proteins upregulated" = proteomic_up)
y =  list("Genes downregulated" = transcriptomic_down, "Proteins downregulated" = proteomic_down)
venn_up <- ggvenn(x, fill_color = c("salmon", "lightgreen"), show_elements = F, text_size = 5, set_name_size = 5)
venn_down <- ggvenn(y, fill_color = c("salmon", "lightgreen"), show_elements = F, text_size = 5, set_name_size = 5)

ggsave("~/R/Figures2/Proteomic_Venn_Acute_Up (7 dpi).png", venn_up, dpi = 1080, height = 7, width = 9)
ggsave("~/R/Figures2/Proteomic_Venn_Acute_Down (7 dpi).png", venn_down, dpi = 1080, height = 7, width = 9)

intersection_up <- intersect(proteomic_up, transcriptomic_up)
paste(intersection_up,"are upregulated at transcriptomic and proteomic level")

intersection_down <- intersect(proteomic_down, transcriptomic_down)
paste(intersection_down,"are downregulated at transcriptomic and proteomic level")
```


```{r}
transcriptomic <- read_excel("~/R/Downloads/DEG_Acute.xlsx")

interest_genes <- readRDS("~/R/Downloads/Genes_Filtered.rds")

transcriptomic <-  transcriptomic %>% 
  filter(Gene %in% interest_genes)

proteomic <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")

# CORRELATION 
transcriptomic <- transcriptomic [, c("Gene", "avg_log2FC","p_val_adj")]

proteomic <- proteomic [, c("Gene", "logFC_Acute_Ctrl", "adj.P.Val_Acute_Ctrl")]

correlation_table <- merge(transcriptomic, proteomic, by = "Gene", all = T)

colnames(correlation_table)[2] <- "log2FC_Transcriptomics"
colnames(correlation_table)[3] <- "p_value_Transcriptomics"
colnames(correlation_table)[4] <- "log2FC_Proteomics"
colnames(correlation_table)[5] <- "p_value_Proteomics"

correlation_table$log2FC_Transcriptomics[is.na(correlation_table$log2FC_Transcriptomics)] <- 0
correlation_table$p_value_Transcriptomics[is.na(correlation_table$p_value_Transcriptomics)] <- 1

correlation_table$log2FC_Proteomics <- as.numeric(correlation_table$log2FC_Proteomics)
correlation_table$p_value_Proteomics <- as.numeric(correlation_table$p_value_Proteomics)

correlation_table <- correlation_table %>% 
  mutate(Tendence = case_when(
    log2FC_Transcriptomics > 0.5 & log2FC_Proteomics > 0.5 ~ "Both upregulated",
    log2FC_Transcriptomics < -0.5 & log2FC_Proteomics < -0.5 ~ "Both downregulated",
    log2FC_Transcriptomics > 0.5 & log2FC_Proteomics < -0.5 ~ "Transcriptomics up, Proteomics down",
    log2FC_Transcriptomics < -0.5 & log2FC_Proteomics > 0.5 ~ "Transcriptomics down, Proteomics up",
    TRUE ~ "Neutral" 
  ))

# P-value Proteomics 0.05
correlation_table1 <- correlation_table %>% 
  filter(p_value_Proteomics <= 1 & p_value_Transcriptomics < 0.05)

correlation <- ggplot(correlation_table1, aes(x = log2FC_Transcriptomics, y = log2FC_Proteomics, label = Gene)) +
  geom_point(aes(color = Tendence), size = 3, alpha = 0.7) +  
  geom_text_repel(aes(color = Tendence), size = 3) +  
  geom_smooth(method = "lm", color = "red", se = FALSE) +  
  theme_minimal() +
  labs(title = "Relationship between Gene Expression and Protein Levels",
       x = "log2FC Transcriptomics",
       y = "log2FC Proteomics") +
  scale_color_manual(values = c(
    "Both upregulated" = "green",
    "Both downregulated" = "red",
    "Transcriptomics up, Proteomics down" = "blue",
    "Transcriptomics down, Proteomics up" = "cyan",
    "Neutral" = "grey10"
  )) +
  theme(legend.position = "right")
correlation
ggsave("~/R/Figures/Proteomic_Correlation_Acute (7 dpi).png", correlation, dpi = 1080, height = 7, width = 9)

```

*SUB-CHRONIC CONDITION (4-6 wpi)*

```{r}
transcriptomic <- read_excel("~/R/Downloads/DEG_Sub-Chronic.xlsx")
proteomic <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")

# VENN DIAGRAMS 
transcriptomic_up <- transcriptomic %>% filter(transcriptomic$avg_log2FC > 0.5 & p_val_adj < 0.05)
transcriptomic_up <- transcriptomic_up$Gene
transcriptomic_down <- transcriptomic %>% filter(transcriptomic$avg_log2FC < -0.5 & p_val_adj < 0.05)
transcriptomic_down <- transcriptomic_down$Gene

proteomic_up <- proteomic %>% filter(expr_Adjpval_Chronic_Ctrl == "up" & as.numeric(logFC_Chronic_Ctrl) > 0.5)
proteomic_up <- proteomic_up$Gene
proteomic_down <- proteomic %>% filter(expr_Adjpval_Chronic_Ctrl == "down" & as.numeric(logFC_Chronic_Ctrl) < -0.5)
proteomic_down <- proteomic_down$Gene

x = list("Genes upregulated" = transcriptomic_up, "Proteins upregulated" = proteomic_up)
y =  list("Genes downregulated" = transcriptomic_down, "Proteins downregulated" = proteomic_down)
venn_up <- ggvenn(x, fill_color = c("salmon", "lightgreen"), show_elements = F, text_size = 5, set_name_size = 5)
venn_down <- ggvenn(y, fill_color = c("salmon", "lightgreen"), show_elements = F, text_size = 5, set_name_size = 5)

ggsave("~/R/Figures2/Proteomic_Venn_Sub-Chronic_Up (4-6 wpi).png", venn_up, dpi = 1080, height = 7, width = 9)
ggsave("~/R/Figures2/Proteomic_Venn_Sub-Chronic_Down (4-6 wpi).png", venn_down, dpi = 1080, height = 7, width = 9)

intersection_up <- intersect(proteomic_up, transcriptomic_up)
paste(intersection_up,"are upregulated at transcriptomic and proteomic level")

intersection_down <- intersect(proteomic_down, transcriptomic_down)
paste(intersection_down,"are downregulated at transcriptomic and proteomic level")

```

```{r}
transcriptomic <- read_excel("~/R/Downloads/DEG_Sub-Chronic.xlsx")

interest_genes <- readRDS("~/R/Downloads/Genes_Filtered.rds")

transcriptomic <-  transcriptomic %>% 
  filter(Gene %in% interest_genes)

proteomic <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")

# CORRELATION
transcriptomic <- transcriptomic [, c("Gene", "avg_log2FC","p_val_adj")]

proteomic <- proteomic [, c("Gene", "logFC_Chronic_Ctrl", "adj.P.Val_Chronic_Ctrl")]

correlation_table <- merge(transcriptomic, proteomic, by = "Gene", all = T)

colnames(correlation_table)[2] <- "log2FC_Transcriptomicsc"
colnames(correlation_table)[3] <- "p_value_Transcriptomics"
colnames(correlation_table)[4] <- "log2FC_Proteomics"
colnames(correlation_table)[5] <- "p_value_Proteomics"

correlation_table$log2FC_Transcriptomics[is.na(correlation_table$log2FC_Transcriptomics)] <- 0
correlation_table$p_value_Transcriptomics[is.na(correlation_table$p_value_Transcriptomics)] <- 1

correlation_table$log2FC_Proteomics <- as.numeric(correlation_table$log2FC_Proteomics)
correlation_table$p_value_Proteomics <- as.numeric(correlation_table$p_value_Proteomics)

correlation_table <- correlation_table %>% 
  mutate(Tendence = case_when(
    log2FC_Transcriptomics > 0.5 & log2FC_Proteomics > 0.5 ~ "Both upregulated",
    log2FC_Transcriptomics < -0.5 & log2FC_Proteomics < -0.5 ~ "Both downregulated",
    log2FC_Transcriptomics > 0.5 & log2FC_Proteomics < -0.5 ~ "Transcriptomics up, Proteomics down",
    log2FC_Transcriptomics < -0.5 & log2FC_Proteomics > 0.5 ~ "Transcriptomics down, Proteomics up",
    TRUE ~ "Neutral"
  ))

# P-value Proteomics 0.05
correlation_table2 <- correlation_table %>% 
  filter(p_value_Proteomics <= 1 & p_value_Transcriptomics < 0.05)

correlation <- ggplot(correlation_table2, aes(x = log2FC_Transcriptomics, y = log2FC_Proteomics, label = Gene)) +
  geom_point(aes(color = Tendence), size = 3, alpha = 0.7) +
  geom_text_repel(aes(color = Tendence), size = 3) + 
  geom_smooth(method = "lm", color = "red", se = FALSE) +  
  theme_minimal() +
  labs(title = "Relationship between Gene Expression and Protein Levels",
       x = "log2FC Transcriptomics",
       y = "log2FC Proteomics") +
  scale_color_manual(values = c(
    "Both upregulated" = "green",
    "Both downregulated" = "red",
    "Transcriptomics up, Proteomics down" = "grey40",
    "Transcriptomics down, Proteomics up" = "grey20",
    "Neutral" = "grey80"
  )) +
  theme(legend.position = "bottom")
correlation
ggsave("~/R/Figures/Proteomic_Correlation_Sub-Chronic (4-6 wpi).png", correlation, dpi = 1080, height = 7, width = 9)
```

```{r}
# Acute
correlation_table1 <- correlation_table1 %>% 
  filter(p_value_Transcriptomics < 0.05)

prefix <- "Acute_"

colnames(correlation_table1) <- sapply(colnames(correlation_table1), function(x) ifelse(x != "Gene", paste0(prefix, x), x))

# Sub-Chronic
correlation_table2 <- correlation_table2 %>% 
  filter(p_value_Transcriptomics < 0.05)

prefix <- "SubChronic_"

colnames(correlation_table2) <- sapply(colnames(correlation_table2), function(x) ifelse(x != "Gene", paste0(prefix, x), x))

# Merge and Data tidy
correlation_table <- merge(correlation_table1, correlation_table2, by = "Gene")

correlation_table <- correlation_table %>% 
  select(matches("Gene|^Acute|^Sub"))

correlation_table <- correlation_table %>% 
  select(matches("Gene|Transcriptomics$|Proteomics$"))

delete_cols <- grep("^\\w+_p_value_\\w+$", colnames(correlation_table), value = TRUE)

correlation_table <- correlation_table[, !colnames(correlation_table) %in% delete_cols]

# Prepare for ggpplot
long_data <- correlation_table %>%
  pivot_longer(cols = -Gene, 
               names_to = c("Condition", "Type"), 
               names_pattern = "(.*)_(.*)",
               values_to = "log2FC")
long_data$Condition <- gsub("_log2FC", "", long_data$Condition)

uninjured_data <- expand.grid(
  Gene = unique(long_data$Gene),
  Condition = "Uninjured",
  Type = unique(long_data$Type),
  log2FC = 0
)
long_data <- bind_rows(long_data, uninjured_data) %>%
  mutate(Condition = factor(Condition, levels = c("Uninjured", "Acute", "SubChronic")))

long_data <- long_data %>%
  mutate(Condition = recode(Condition,
                            "Acute" = "7 dpi",
                            "SubChronic" = "4-6 wpi",
                            "Uninjured" = "Before-Injury"))

saveRDS(long_data, file = "~/R/Downloads/Data_Omics.rds")

layout_proteomica <- ggplot(long_data, aes(x = Condition, y = log2FC, color = Type, group = Type)) +
  geom_line() +
  geom_point() + 
  facet_wrap(~ Gene, scales = "free_y") +
  labs(
    title = "Proteomics vs Transcriptomics Comparison",
    x = "Condition", 
    y = "log2 Fold Change",
    color = "Data Type"
  ) +
  theme_minimal(base_family = "sans") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text.y = element_text(size = 5),
    legend.position = "right"
  )

layout_proteomica

ggsave("~/R/Figures2/Layout Proteomics-Transcriptomics.png", layout_proteomica, dpi = 1080, height = 7, width = 9)


# Classification by correlation at different times
wide_data <- long_data %>%
  pivot_wider(names_from = Type, values_from = log2FC) %>%
  rename(
    transcript = Transcriptomics,
    protein = Proteomics
  ) %>%
  mutate(
    sign_match = case_when(
      transcript > 0 & protein > 0 ~ "pos",
      transcript < 0 & protein < 0 ~ "neg",
      TRUE ~ "no"
    )
  )

gene_class <- wide_data %>%
  select(Gene, Condition, sign_match) %>%
  pivot_wider(names_from = Condition, values_from = sign_match) %>%
  mutate(
    correlation_group = case_when(
      `7 dpi` %in% c("pos", "neg") & `4-6 wpi` %in% c("pos", "neg") ~ "correlation in both",
      `7 dpi` %in% c("pos", "neg") & `4-6 wpi` == "no" ~ "correlation in 7 dpi",
      `7 dpi` == "no" & `4-6 wpi` %in% c("pos", "neg") ~ "correlation in 4-6 wpi",
      TRUE ~ "no correlation"
    )
  ) %>%
  select(Gene, correlation_group)

# Add classification
long_data_with_group <- long_data %>%
  left_join(gene_class, by = "Gene")

# No
long_data_no <- long_data_with_group %>%
  filter(correlation_group == "no correlation")

# Acute
long_data_acute <- long_data_with_group %>%
  filter(correlation_group == "correlation in 7 dpi")

long_data_acute <- long_data_acute %>%
  filter(!(Gene == "Hapln2"))

# Chronic
long_data_chronic <- long_data_with_group %>%
  filter(correlation_group == "correlation in 4-6 wpi")

# Both
long_data_both <- long_data_with_group %>%
  filter(correlation_group == "correlation in both" | Gene == "Hapln2")

long_data_both <- long_data_both %>%
  filter(!(Gene %in% c("C1qa", "F2", "Lama5")))

long_data_both_x <- long_data_with_group %>%
  filter(Gene %in% c("C1qa", "F2", "Lama5"))


# Acute
layout_proteomica_acute <- ggplot(long_data_acute, aes(x = Condition, y = log2FC, color = Type, group = Type)) +
  geom_line() +
  geom_point() + 
  facet_wrap(~ Gene, scales = "free_y") +
  labs(
    title = "Proteomics vs Transcriptomics Comparison",
    x = "Condition", 
    y = "log2 Fold Change",
    color = "Data Type"
  ) +
  theme_minimal(base_family = "sans") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text.y = element_text(size = 5),
    legend.position = "right"
  )

layout_proteomica_acute

ggsave("~/R/Figures2/Layout Proteomics-Transcriptomics_Acute.png", layout_proteomica_acute, dpi = 1080, height = 7, width = 9)

# Both
layout_proteomica_both <- ggplot(long_data_both, aes(x = Condition, y = log2FC, color = Type, group = Type)) +
  geom_line() +
  geom_point() + 
  facet_wrap(~ Gene, scales = "free_y") +
  labs(
    title = "Proteomics vs Transcriptomics Comparison",
    x = "Condition", 
    y = "log2 Fold Change",
    color = "Data Type"
  ) +
  theme_minimal(base_family = "sans") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text.y = element_text(size = 5),
    legend.position = "right"
  )

layout_proteomica_both

ggsave("~/R/Figures2/Layout Proteomics-Transcriptomics_Both.png", layout_proteomica_both, dpi = 1080, height = 7, width = 9)

# Both
layout_proteomica_both_x <- ggplot(long_data_both_x, aes(x = Condition, y = log2FC, color = Type, group = Type)) +
  geom_line() +
  geom_point() + 
  facet_wrap(~ Gene, scales = "free_y") +
  labs(
    title = "Proteomics vs Transcriptomics Comparison",
    x = "Condition", 
    y = "log2 Fold Change",
    color = "Data Type"
  ) +
  theme_minimal(base_family = "sans") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text.y = element_text(size = 5),
    legend.position = "right"
  )

layout_proteomica_both_x

ggsave("~/R/Figures2/Layout Proteomics-Transcriptomics_Both.png", layout_proteomica_both_x, dpi = 1080, height = 7, width = 9)

# Chronic
layout_proteomica_chronic <- ggplot(long_data_chronic, aes(x = Condition, y = log2FC, color = Type, group = Type)) +
  geom_line() +
  geom_point() + 
  facet_wrap(~ Gene, scales = "free_y") +
  labs(
    title = "Proteomics vs Transcriptomics Comparison",
    x = "Condition", 
    y = "log2 Fold Change",
    color = "Data Type"
  ) +
  theme_minimal(base_family = "sans") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text.y = element_text(size = 5),
    legend.position = "right"
  )

layout_proteomica_chronic

ggsave("~/R/Figures2/Layout Proteomics-Transcriptomics_Chronic.png", layout_proteomica_chronic, dpi = 1080, height = 7, width = 9)

# No correlation
layout_proteomica_no <- ggplot(long_data_no, aes(x = Condition, y = log2FC, color = Type, group = Type)) +
  geom_line() +
  geom_point() + 
  facet_wrap(~ Gene, scales = "free_y") +
  labs(
    title = "Proteomics vs Transcriptomics Comparison",
    x = "Condition", 
    y = "log2 Fold Change",
    color = "Data Type"
  ) +
  theme_minimal(base_family = "sans") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text.y = element_text(size = 5),
    legend.position = "right"
  )

layout_proteomica_no

ggsave("~/R/Figures2/Layout Proteomics-Transcriptomics_No.png", layout_proteomica_no, dpi = 1080, height = 7, width = 9)

genes_corr <- unique(long_data_both$Gene)
saveRDS(genes_corr, file = "~/R/Downloads/Genes_Correlation.rds")

```

