---
title: "Meta-analysis"
output: html_document
date: "2025-02-26"
---
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

Packages used: 

```{r}
library(Seurat)
library(Matrix)
library(biomaRt)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(ggplot2)
library(dplyr)
library(stringr)
library(glmGamPoi)
library(readxl)
library(EnhancedVolcano)
library(BiocParallel)
library(speckle)
library(presto)
library(enrichplot)
library(clusterProfiler)
library(MAST)
library(plotly)
library(openxlsx)
library(patchwork)
library(tibble)
```

1. First of all, we need to prepare the three datasets for integration. One of the most important points is standarize the labels of the dataset.The different datasets are already filtered by QC metrics. (% mitochondrial < 0.2, no doublets)


Dataset Milich 

```{r}
matrix <- readMM("~/R/Data/Dataset1.mtx")
metadata <- read.delim("~/R/Data/metadata1.tsv", header = T)
features <- read.delim("~/R/Data/features1.tsv")
barcodes <- rownames(metadata)
```

```{r}
metadata <- metadata %>% dplyr::select(nCount_RNA, nFeature_RNA, time, celltype, sample_id)

metadata$technology <- "scRNA"
metadata$dataset <- "Milich_2021"

rownames(matrix) <- features$mgi_symbol
colnames(matrix) <- rownames(barcodes)
```

```{r}
seurat <- CreateSeuratObject(matrix,assay = "RNA", project = "Milich_2021", meta.data = metadata, min.cells = 10, min.features = 100)

remove_features <- nrow(matrix) - nrow(seurat)
remove_cells <- ncol(matrix) - ncol(seurat)

seurat <- subset(seurat, subset = time %in% c("Uninjured", "1dpi", "7dpi"))

remove_features_f <- nrow(matrix) - nrow(seurat)
remove_cells_f <- ncol(matrix) - nrow(seurat)

# Remove replicates partially enriched
# seurat <- subset(seurat, subset = sample_id %in% c("1dpi_sample1", "7dpi_sample1", "uninj_sample1"))

rm(matrix,barcodes,metadata,features)
```

Dataset Matson 2022

```{r}
matrix2 <- readRDS("~/R/Data/Dataset2.rds")
```

```{r}
counts2 <- matrix2@assays[["RNA"]]@counts

metadata2 <- matrix2@meta.data

metadata2$orig.ident <- "Matson_2022 (Single cell Atlas)"

colnames(metadata2)[colnames(metadata2) == "coarse_clusters"] <- "celltype"
colnames(metadata2)[colnames(metadata2) == "condition"] <- "time"
colnames(metadata2)[colnames(metadata2) == "nGene"] <- "nFeature_RNA"
colnames(metadata2)[colnames(metadata2) == "nUMI"] <- "nCount_RNA"

metadata2$time <- gsub("C_1wpi", "7dpi", metadata2$time)
metadata2$time <- gsub("A_Uninj", "Uninjured", metadata2$time)
metadata2$time <- gsub("B_1dpi", "1dpi", metadata2$time)

colnames(metadata2) <- make.unique(colnames(metadata2))

metadata2 <- metadata2 %>% select(matches("nCount_RNA|nFeature_RNA|time|celltype"))

metadata2 <- metadata2 [, -c(3,4)]

metadata2$technology <- "snRNA"
metadata2$dataset <- "Matson_2022"

```

```{r}
seurat2 <- CreateSeuratObject(counts2, assay = "RNA", project = "Matson_2022 (Single cell Atlas)", meta.data = metadata2, min.cells = 10, min.features = 100)

remove_features2 <- nrow(matrix2) - nrow(seurat2)
remove_cells2 <- ncol(matrix2) - ncol(seurat2)

seurat2 <- subset(seurat2, subset = time %in% c("Uninjured", "1dpi", "7dpi"))

remove_features2_f <- nrow(matrix2) - nrow(seurat2)
remove_cells2_f <- ncol(matrix2) - ncol(seurat2)

rm(matrix2, counts2, metadata2)
```

Dataset Skinnider_2024

```{r}

matrix3 <- readMM("~/R/Data/Dataset3.mtx")
gc()
barcodes3 <- read.delim("~/R/Data/barcodes3.txt.gz", header = F)
metadata3 <- read.delim("~/R/Data/metadata3.txt", header = T)
features3 <- read.delim("~/R/Data/features3.txt.gz", header = F)

```


```{r}

colnames(matrix3) <- barcodes3$V1
rownames(matrix3) <- features3$V1

rownames(metadata3) <- metadata3$barcode

matrix3 <- as(matrix3, "dgCMatrix")
gc()

metadata3 <- metadata3 %>% rename(time = label)

metadata3$time <- recode(metadata3$time, 
                         "contusion" = "7dpi", 
                         "uninjured" = "Uninjured")

metadata3 <- metadata3 %>% rename(celltype = layer2)

metadata3 <- metadata3 %>% select(matches("nCount_RNA|nFeature_RNA|time|celltype"))

metadata3$technology <- "snRNA"
metadata3$dataset <- "Skinnider_2024"

```

```{r}
seurat3  <- CreateSeuratObject(matrix3, assay = "RNA", project = "Skinnider_2024 (Tabulae Paralytica)",  meta.data = metadata3, min.cells = 10, min.features = 100)

remove_features3 <- nrow(matrix3) - nrow(seurat3)
remove_cells3 <- ncol(matrix3) - ncol(seurat3)

seurat3 <- subset(seurat3, subset = time %in% c("Uninjured", "7dpi"))

remove_features3_f <- nrow(matrix3) - nrow(seurat3)
remove_cells3_f <- ncol(matrix3) - ncol(seurat3)

rm(metadata3, matrix3, barcodes3, features3, filtered_matrix)
```


Quality control check

Even though methods of the papers indicate that the matrices have already been filtered, we make a small QC check.

```{r}
# Quality control of features
print(paste("Features removed for QC:", remove_features + remove_features2 + remove_features3))
paste("Features:", nrow(seurat) + nrow(seurat2) + nrow(seurat3))

# Quality control of cells
print(paste("Cells removed for QC:", remove_cells + remove_cells2 + remove_cells3))
print(paste("Cells of other experimental conditionns:", (remove_cells_f + remove_cells2_f + remove_cells3_f) - (remove_cells + remove_cells2 + remove_cells3)))
paste("Cells:", ncol(seurat) + ncol(seurat2) + ncol(seurat3))

```

Dataset - Kathe 2022

```{r}

matrix5 <- readMM("~/R/Data/Dataset5.mtx")
barcodes5 <- read.table("~/R/Data/barcodes5.txt.gz", header = F)
features5 <- read.delim("~/R/Data/features5.txt.gz", header = F)
metadata5 <- read.delim("~/R/Data/metadata5.txt")


colnames(matrix5) <- barcodes5$V1
rownames(matrix5) <- features5$V1

rownames(metadata5) <- metadata5$barcode

metadata5 <- metadata5 %>% rename(time = label)
metadata5 <- metadata5 %>% rename(celltype = global_cell_type)

metadata5$time <- recode(metadata5$time, "SCI" = "Sub-Chronic")

metadata5 <- metadata5 %>% select(matches("time|celltype"))

metadata5$technology <- "snRNA"
metadata5$dataset <- "Kathe_2022"

```


```{r}
seurat5 <- CreateSeuratObject(matrix5, assay = "RNA", project = "Kathe_2022",  meta.data = metadata5, min.cells = 50, min.features = 100)

remove_features5 <- nrow(matrix5) - nrow(seurat5)
remove_cells5 <- ncol(matrix5) - ncol(seurat5)

seurat5 <- subset(seurat5, subset = time %in% c("Uninjured", "Sub-Chronic"))

remove_cells5_f <- ncol(matrix5) - ncol(seurat5)

rm(matrix5, barcodes5, metadata5, features5)

```

```{r}
save.image(file = "~/R/Data/Seurats_Acute.RData")
```

Normalization

```{r}
# Normalization of 3 datasets
Idents(seurat) <- "celltype"

seurat_list <- list(seurat, seurat2, seurat3, seurat5)
seurat_list_norm <- list()

for (i in seq_along(seurat_list)) {
  seurat_norm_i <- NormalizeData(seurat_list[[i]], normalization.method = "LogNormalize", verbose = TRUE)
  seurat_norm_i <- FindVariableFeatures(seurat_norm_i, selection.method = "vst", nfeatures = 5000, verbose = TRUE)
  seurat_list_norm[[i]] <- seurat_norm_i
  rm(seurat_norm_i)
}

seurat_norm <- seurat_list_norm[[1]]
seurat2_norm <- seurat_list_norm[[2]]
seurat3_norm <- seurat_list_norm[[3]]
seurat5_norm <- seurat_list_norm[[4]]
```

```{r}
rm(seurat_list_norm, seurat, seurat2, seurat3)
```


Transfer cell type annotations

FindTransferAnchors, TransferData and AddMetadata are used to update the seurat objects with the new predicted annotations based on reference annotations provided by Seurat from Milich 2021.

```{r}

# Cell Annotation with Seurat from Milich as reference
# Dataset 1
anchors <- FindTransferAnchors(reference = seurat5_norm, query = seurat_norm, dims = 1:20)

predictions <- TransferData(anchorset = anchors, refdata = seurat5_norm$celltype, dims = 1:20)

seurat_norm <- AddMetaData(seurat_norm, metadata = predictions)

# Dataset 2
anchors <- FindTransferAnchors(reference = seurat5_norm, query = seurat2_norm, dims = 1:20)

predictions <- TransferData(anchorset = anchors, refdata = seurat5_norm$celltype, dims = 1:20)

seurat2_norm <- AddMetaData(seurat2_norm, metadata = predictions)

# Dataset 3
anchors <- FindTransferAnchors(reference = seurat5_norm, query = seurat3_norm, dims = 1:20)

predictions <- TransferData(anchorset = anchors, refdata = seurat5_norm$celltype, dims = 1:20)

seurat3_norm <- AddMetaData(seurat3_norm, metadata = predictions)

rm(seurat5_norm, seurat5)

```

Scale, RunPCA, Run UMAP

# seurat_norm borrat
```{r}

seurat_list <- list(seurat_norm, seurat2_norm, seurat3_norm)

features <- SelectIntegrationFeatures(
  object.list = seurat_list, 
  nfeatures = 3500)

seurat_list <- lapply(X = seurat_list, FUN = function(x) {x <- ScaleData(x, features =  features)
x <- RunPCA(x, features = features, npcs = 30)
x <- RunUMAP(x, dims = 1:20)})

seurat_norm <- seurat_list[[1]]
seurat2_norm <- seurat_list[[2]]
seurat3_norm <- seurat_list[[3]]

```

UMAP of the 3 datasets before integration

```{r}
umap <- DimPlot(seurat_norm, split.by = "dataset", group.by = "predicted.id", label = T, label.size = 3) + theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 7)) + labs(x = "UMAP 1", y = "UMAP 2")

umap2 <- DimPlot(seurat2_norm, split.by = "dataset", group.by = "predicted.id", label = T, label.size = 3) + theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 7)) + labs(x = "UMAP 1", y = "UMAP 2")

umap3 <- DimPlot(seurat3_norm, split.by = "dataset", group.by = "predicted.id", label = T, label.size = 3) + theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 7)) + labs(x = "UMAP 1", y = "UMAP 2")

umap + umap2 + umap3

ggsave("~/R/Figures/UMAP_BeforeIntegration_Acute.png", umap + umap2 + umap3, dpi = 1080, height = 14, width = 8)

rm(umap, umap2, umap3)
rm(seurat2_norm, seurat_norm, seurat3_norm)
```

Integrate the three datasets and scale the integrated seurat object.

```{r}
ecm <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")
genes_of_interest <- ecm$Gene

features <- c(features, genes_of_interest)

features <- unique(features)

features <- intersect(features, rownames(seurat_list[[1]]))
features <- intersect(features, rownames(seurat_list[[2]])) 
features <- intersect(features, rownames(seurat_list[[3]]))

gc()
anchors_integration <- FindIntegrationAnchors(object.list = seurat_list, dims = 1:30, verbose = TRUE, reduction = "rpca", anchor.features = features, scale = F, normalization.method = "LogNormalize", k.anchor = 50)

gc()
seurat_integrated <- IntegrateData(anchorset = anchors_integration, dims = 1:30, verbose = TRUE, normalization.method = "LogNormalize") 

rm(seurat_list, anchors, anchors_integration)

save.image(file = "~/R/Downloads/Acute_Data_RPCA.RData")
```

* Correct metadata issues

```{r}

metadata <- seurat_integrated@meta.data
metadata <- metadata %>% select(matches("nCount_RNA|nFeature_RNA|time|celltype|technology|dataset|predicted.id"))

metadata$celltype <- NULL
colnames(metadata)[colnames(metadata) == "predicted.id"] <- "celltype"

# Return check metadata into seurat object
seurat_integrated@meta.data <- metadata

```


```{r}
# Descriptive
table(seurat_integrated$celltype)
table(seurat_integrated$time)
table(seurat_integrated$dataset)
``` 

Cell Cycle Score

```{r}

cell_cycle_markers <- read.csv("~/R/Data/Mus_musculus.csv")
s_genes <- cell_cycle_markers %>% dplyr::filter(phase == "S") %>% pull("geneID")
g2m_genes <- cell_cycle_markers %>% dplyr::filter(phase == "G2/M") %>% pull("geneID")
s_genes <- AnnotationDbi::select(org.Mm.eg.db,
                                   keys = s_genes,
                                   keytype = "ENSEMBL",
                                   columns = c("SYMBOL"))
g2m_genes <- AnnotationDbi::select(org.Mm.eg.db,
                                   keys = g2m_genes,
                                   keytype = "ENSEMBL",
                                   columns = c("SYMBOL"))

seurat_integrated <- CellCycleScoring(seurat_integrated, g2m.features = g2m_genes$SYMBOL, s.features = s_genes$SYMBOL)

```

UMAP after integration

```{r}
seurat_integrated <- ScaleData(seurat_integrated)
seurat_integrated <- RunPCA(seurat_integrated, verbose = T, dims = 1:20) 
seurat_integrated <- RunUMAP(seurat_integrated, dims = 1:20)

umap <- DimPlot(seurat_integrated, split.by = "dataset", group.by = "celltype", label = T, label.size = 3) + theme(axis.text.x = element_text(angle = 60, hjust = 1, size = 9)) + labs(x = "UMAP 1", y = "UMAP 2")
umap
  ggsave("~/R/Figures/UMAP_AfterIntegration_Acute.png", umap, dpi = 1080, height = 7, width = 12)

celltype <- DimPlot(seurat_integrated, group.by = "celltype") + ggtitle("Celltypes")
celltype

dataset <- DimPlot(seurat_integrated, group.by = "dataset") + ggtitle("Datasets")
dataset

ggsave("~/R/Figures/Cells_Acute.png", celltype, dpi = 1080, height = 7, width = 10)
ggsave("~/R/Figures/Dataset_Acute.png", dataset, dpi = 1080, height = 7, width = 10)
```

Possible covariates

```{r}
cc_score <- DimPlot(seurat_integrated,
        reduction = "pca",
        group.by= "Phase",
        split.by = "time")
cc_score
ggsave("~/R/Figures/CC_Acute.png", cc_score, dpi = 1080, height = 7, width = 9)

# Differences between cells in different CC Phase

pca_dataset <- DimPlot(seurat_integrated,
        reduction = "pca",
        group.by= "dataset")
pca_dataset

ggsave("~/R/Figures/PCA_Dataset_Sub-Chronic.png", pca_dataset, dpi = 1080, height = 7, width = 9)
```

LISI

```{r}
library(lisi)

embeddings <- Embeddings(seurat_integrated, reduction = "umap") 

metadata <- seurat_integrated@meta.data
batch_labels <- metadata$dataset  
celltype_labels <- metadata$celltype

lisi_scores <- lisi::compute_lisi(X = embeddings, meta_data = data.frame(batch = seurat_integrated$dataset, celltype = seurat_integrated$celltype), label_colnames = c("batch", "celltype"), perplexity = 100
)

lisi_df <- data.frame(
  Batch_LISI = lisi_scores$batch,
  CellType_LISI = lisi_scores$celltype
)

batch_plot <- ggplot(lisi_df, aes(x = Batch_LISI)) +
  geom_density(fill = "salmon", alpha = 0.7) +
  geom_vline(xintercept = 1, linetype = "dashed", color = "red") +
  labs(title = "LISI - Batch (Dataset)",
       x = "LISI Score",
       y = "Density") +
  theme_minimal() +
  xlim(c(1, max(lisi_df$Batch_LISI)))

celltype_plot <- ggplot(lisi_df, aes(x = CellType_LISI)) +
  geom_density(fill = "steelblue", alpha = 0.7) +
  labs(title = "LISI - Cell Type",
       x = "LISI Score",
       y = "Density") +
  theme_minimal() +
  xlim(c(1, max(lisi_df$CellType_LISI)))

combined_plot <- batch_plot + celltype_plot +
  plot_annotation(title = "Evaluation of the Integration by LISI",
                  subtitle = paste("Número de células:", nrow(embeddings)))

print(combined_plot)

summary_stats <- data.frame(
  Metric = c("Batch LISI", "CellType LISI"),
  Mean = c(mean(lisi_df$Batch_LISI), mean(lisi_df$CellType_LISI)),
  Median = c(median(lisi_df$Batch_LISI), median(lisi_df$CellType_LISI)),
  SD = c(sd(lisi_df$Batch_LISI), sd(lisi_df$CellType_LISI))
)

print(summary_stats)

```

```{r}

seurat_integrated@meta.data$celltype <- recode(seurat_integrated@meta.data$celltype, "Endothelial cells" = "Endothelial")

seurat_integrated$celltype <- recode(seurat_integrated$celltype, "Endothelial cells" = "Endothelial")

seurat_integrated@meta.data$`Endothelial cells` -> seurat_integrated@meta.data$Endothelial


```

Filter ECM genes

```{r}
proteins.ME <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")
genes_to_keep <- proteins.ME$Gene
rownames_seurat <- rownames(seurat_integrated)
rows_to_keep <- which(rownames_seurat %in% genes_to_keep)
seurat_integrated_ME <- seurat_integrated[rows_to_keep, ]

paste("ECM Genes present in the dataset:", nrow(seurat_integrated_ME))

```

Differential expression analysis.

FindMarkers analysis

```{r}

adjust <- c("S.Score","G2M.Score")

Idents(seurat_integrated_ME) <- "time"

DEG <- FindMarkers(seurat_integrated_ME,  ident.1 = "7dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.02, test.use = "MAST", layer = "data", latent.vars = adjust)

DEG$Gene <- rownames(DEG)

DEG$p_val_adj <- ifelse(DEG$p_val_adj == 0, 1e-305, DEG$p_val_adj)

volcanoplot <- EnhancedVolcano(DEG, lab = rownames(DEG), x = "avg_log2FC", y = "p_val_adj", pCutoff = 0.05, FCcutoff = 0.75, title = "Differential expression between Uninjured and Acute conditions", drawConnectors = T, labSize = 3)

write.xlsx(DEG, "~/R/Data/DEG_Acute.xlsx")

ggsave("~/R/Figures/Volcano_Acute.png", volcanoplot, dpi = 1080, height = 7, width = 9)

##############################################################################################################

DEG2 <- FindMarkers(seurat_integrated_ME,  ident.1 = "1dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.02, test.use = "MAST", layer = "data",  latent.vars = adjust)

DEG2$Gene <- rownames(DEG2)

DEG2$p_val_adj <- ifelse(DEG2$p_val_adj == 0, 1e-305, DEG2$p_val_adj)

volcanoplot2 <- EnhancedVolcano(DEG2, lab = rownames(DEG2), x = "avg_log2FC", y = "p_val_adj", pCutoff = 0.05, FCcutoff = 0.75, title = "Differential expression between Uninjured and Acute conditions", drawConnectors = T, labSize = 3)

ggsave("~/R/Figures/Volcano_1dpi.png", volcanoplot2, dpi = 1080, height = 7, width = 9)

write.xlsx(DEG2, "~/R/Data/DEG_FirstDay.xlsx")

```

Volcano plot 3-4

```{r}
DEG_matrisome <- merge(DEG, proteins.ME[, c("Gene", "Matrisome.Category")], by = "Gene", all.x = TRUE)

volcano_plot3 <- ggplot(DEG_matrisome, aes(x = avg_log2FC, y = -log10(p_val_adj), color = Matrisome.Category)) +
  geom_point(size = 3, alpha = 0.7) +
  scale_color_viridis_d() + 
  geom_text_repel(
    aes(
      label = ifelse(p_val_adj < 0.05 & abs(avg_log2FC) > 1, as.character(Gene), "")
    ),
    size = 3,
    alpha = 0.8,
    box.padding = unit(0.35, "lines"),  
    point.padding = unit(0.3, "lines"),  
    max.overlaps = Inf,  
    show.legend = FALSE  
  ) +
  theme_minimal() +
  labs(
    title = "Differential Expression in SCI",
    x = "log2 Fold Change", 
    y = "-log10 Adjusted p-value",
    color = "Matrisome Category" 
  ) +
  theme(
    legend.position = "right",
    plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),
    axis.title = element_text(size = 12),
    axis.text = element_text(size = 10),
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 10)
  ) +
  geom_vline(xintercept = c(-0.75, 0.75), linetype = "dashed", color = "gray50") +  
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "gray50")

volcano_plot3

ggsave("~/R/Figures/Volcano_ECM_Acute.png", volcano_plot3, dpi = 1080, height = 7, width = 9)
```

7dpi

Enrichment analysis

```{r}

relevant_genes <- subset(DEG, p_val_adj < 0.05 & abs(avg_log2FC) > 0.5)

relevant_genes <- rownames(relevant_genes)

deg_entrez <- bitr(relevant_genes, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)
deg_entrez <- deg_entrez$ENTREZID

# GO ENRICHMENT
go_enrich <- enrichGO(
  gene = deg_entrez,
  OrgDb = org.Mm.eg.db,
  keyType = "ENTREZID", 
  ont = "BP", qvalueCutoff =  0.1,
  pAdjustMethod = "fdr")

go_terms <- go_enrich@result[["Description"]]

keywords <- c("neuro", "inflammation", "Wnt", "Rho", "regeneration", "remodelling", "SCI", "axonal", "axon","BMP", "spinal")

terms_of_interest <- go_terms[sapply(go_terms, function(x) any(grepl(paste(keywords, collapse="|"), x, ignore.case = TRUE)))]

terms_of_interest <- terms_of_interest[!grepl("muscle", terms_of_interest, ignore.case = TRUE)]

go_enrich_df <- as.data.frame(go_enrich@result)

filtered_go_enrich_df <- go_enrich_df %>% filter(Description %in% terms_of_interest)

filtered_go_enrich <- go_enrich
filtered_go_enrich@result <- filtered_go_enrich_df  

go_enrich_barplot <- barplot(filtered_go_enrich) + 
  theme(axis.text.y = element_text(size = 9)) + 
  ggtitle("GO ENRICHMENT ANALYSIS OF DEG - 7dpi") + labs(caption = "GO Terms Filtered by Spinal Cord Injury related Biological Process")
go_enrich_barplot

ggsave("~/R/Figures/Enrichment_Analysis_Acute.png", go_enrich_barplot, dpi = 1080, height = 7, width = 9)
```

For different cell types 

Dotplot

```{r}

DEG$baseline_exp <- 0

DEG_upregulated <- DEG %>%
  arrange(desc(avg_log2FC)) %>% head(15)

DEG_downregulated <- DEG %>%
  arrange(desc(avg_log2FC)) %>% tail(15)

DEG_TOP <- bind_rows(DEG_downregulated, DEG_upregulated)

# DOTPLOT
genes_interest <- c(DEG_upregulated$Gene, DEG_downregulated$Gene)

seurat_integrated_ME@meta.data$time <- factor(seurat_integrated_ME@meta.data$time)
seurat_integrated_ME@meta.data$celltype <- factor(seurat_integrated_ME@meta.data$celltype)

celltype_DEG <- DotPlot(seurat_integrated_ME, features = unique(genes_interest), group.by = "celltype", split.by = "time", dot.scale = 5, cols = c("red","green", "blue")) + theme_minimal() + labs(title = "Differential Expression of Genes by Time and Cell Type") + theme(axis.text.x = element_text(angle = 45, hjust = 1))
celltype_DEG

ggsave("~/R/Figures/DotPlot_Acute.png", celltype_DEG, dpi = 1080, height = 10, width = 9)

```

HeatMap

```{r}
Idents(seurat_integrated_ME) <- "time"

astro <- subset(seurat_integrated_ME, subset = celltype == "Astrocytes")
DEG_astro <- FindMarkers(astro, ident.1 = "7dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_astro$Gene <- rownames(DEG_astro)
DEG_astro$celltype <- "Astrocytes"

neuron <- subset(seurat_integrated_ME, subset = celltype == "Neurons")
DEG_neuron <- FindMarkers(neuron, ident.1 = "7dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_neuron$Gene <- rownames(DEG_neuron)
DEG_neuron$celltype <- "Neurons"

microglia <- subset(seurat_integrated_ME, subset = celltype == "Microglia")
DEG_microglia <- FindMarkers(microglia, ident.1 = "7dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_microglia$Gene <- rownames(DEG_microglia)
DEG_microglia$celltype <- "Microglia"

opc <- subset(seurat_integrated_ME, subset = celltype == "OPCs")
DEG_opc <- FindMarkers(opc, ident.1 = "7dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_opc$Gene <- rownames(DEG_opc)
DEG_opc$celltype <- "OPCs"

endothelial <- subset(seurat_integrated_ME, subset = celltype == "Endothelial")
DEG_endothelial <- FindMarkers(endothelial, ident.1 = "7dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_endothelial$Gene <- rownames(DEG_endothelial)
DEG_endothelial$celltype <- "Endothelial"

oligodendrocyte <- subset(seurat_integrated_ME, subset = celltype == "Oligodendrocytes")
DEG_oligodendrocyte <- FindMarkers(oligodendrocyte, ident.1 = "7dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_oligodendrocyte$Gene <- rownames(DEG_oligodendrocyte)
DEG_oligodendrocyte$celltype <- "Oligodendrocytes"

combined_results <- bind_rows(DEG_astro, DEG_neuron, DEG_microglia, DEG_opc, DEG_oligodendrocyte, DEG_endothelial)

datasets <- list(DEG_astro, DEG_neuron, DEG_microglia, DEG_opc, DEG_oligodendrocyte, DEG_endothelial)

top_gene_names <- unlist(lapply(datasets, function(df) {
  df_sorted <- df[order(df$avg_log2FC, decreasing = TRUE), ]
  top2_up_genes <- head(df_sorted$Gene, 2)
  top2_down_genes <- tail(df_sorted$Gene, 2)
  c(top2_up_genes, top2_down_genes)
}))

genes_up <- DEG_upregulated$Gene
genes_down <- DEG_downregulated$Gene

combined_results <- combined_results %>%
  filter(Gene %in% genes_down | Gene %in% genes_up | Gene %in% top_gene_names)

heatmap_data <- combined_results %>%
  select(Gene, celltype, avg_log2FC) %>%
  pivot_wider(names_from = celltype, values_from = avg_log2FC) %>%
  column_to_rownames("Gene") %>%
  as.matrix()

heatmap_data[is.na(heatmap_data)] <- 0

library(ComplexHeatmap)
heatmap <- Heatmap(
  heatmap_data,
  col = circlize::colorRamp2(c(-7, 0, 7), c("blue", "white", "red")),
  row_names_gp = gpar(fontsize = 8),
  column_names_gp = gpar(fontsize = 8),
  name = "log2FC", row_title = "Genes", row_title_side = "right", column_title = "Cell types", column_title_side = "bottom"
)

heatmap

```

For 1 dpi

Enrichment analysis

```{r}

relevant_genes2 <- subset(DEG2, p_val_adj < 0.05 & abs(avg_log2FC) > 0.5)

relevant_genes2 <- rownames(relevant_genes2)

deg_entrez2 <- bitr(relevant_genes2, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)
deg_entrez2 <- deg_entrez2$ENTREZID

# GO ENRICHMENT
go_enrich2 <- enrichGO(
  gene = deg_entrez2,
  OrgDb = org.Mm.eg.db,
  keyType = "ENTREZID", 
  ont = "BP", qvalueCutoff =  0.1,
  pAdjustMethod = "fdr")

go_terms2 <- go_enrich2@result[["Description"]]

keywords <- c("neuro", "inflammation", "Wnt", "Rho", "regeneration", "remodelling", "SCI", "axonal", "axon","BMP", "spinal")

terms_of_interest <- go_terms2[sapply(go_terms2, function(x) any(grepl(paste(keywords, collapse="|"), x, ignore.case = TRUE)))]

terms_of_interest <- terms_of_interest[!grepl("muscle", terms_of_interest, ignore.case = TRUE)]

go_enrich_df2 <- as.data.frame(go_enrich2@result)

filtered_go_enrich_df2 <- go_enrich_df2 %>% filter(Description %in% terms_of_interest)

filtered_go_enrich2 <- go_enrich2
filtered_go_enrich2@result <- filtered_go_enrich_df2  

go_enrich_barplot2 <- barplot(filtered_go_enrich2) + 
  theme(axis.text.y = element_text(size = 9)) + 
  ggtitle("GO ENRICHMENT ANALYSIS OF DEG - 1dpi") + labs(caption = "GO Terms Filtered by Spinal Cord Injury related Biological Process")

go_enrich_barplot2

ggsave("~/R/Figures/Enrichment_Analysis_FirstDay.png", go_enrich_barplot2, dpi = 1080, height = 7, width = 9)
```



```{r}

DEG2$baseline_exp <- 0

DEG_upregulated2 <- DEG2 %>%
  arrange(desc(avg_log2FC)) %>% head(15)

DEG_downregulated2 <- DEG2 %>%
  arrange(desc(avg_log2FC)) %>% tail(15)

DEG_TOP2 <- bind_rows(DEG_downregulated2, DEG_upregulated2)

```

HeatMap

```{r}
Idents(seurat_integrated_ME) <- "time"

DEG_astro2 <- FindMarkers(astro, ident.1 = "1dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_astro2$Gene <- rownames(DEG_astro2)
DEG_astro2$celltype <- "Astrocytes"

DEG_neuron2 <- FindMarkers(neuron, ident.1 = "1dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_neuron2$Gene <- rownames(DEG_neuron2)
DEG_neuron2$celltype <- "Neurons"

DEG_microglia2 <- FindMarkers(microglia, ident.1 = "1dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_microglia2$Gene <- rownames(DEG_microglia2)
DEG_microglia2$celltype <- "Microglia"

DEG_opc2 <- FindMarkers(opc, ident.1 = "1dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_opc2$Gene <- rownames(DEG_opc2)
DEG_opc2$celltype <- "OPCs"

DEG_endothelial2 <- FindMarkers(endothelial, ident.1 = "1dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_endothelial2$Gene <- rownames(DEG_endothelial2)
DEG_endothelial2$celltype <- "Endothelial"

DEG_oligodendrocyte2 <- FindMarkers(oligodendrocyte, ident.1 = "1dpi", ident.2 = "Uninjured", logfc.threshold = 0, min.pct = 0.05, test.use = "MAST")
DEG_oligodendrocyte2$Gene <- rownames(DEG_oligodendrocyte2)
DEG_oligodendrocyte2$celltype <- "Oligodendrocytes"

combined_results2 <- bind_rows(DEG_astro2, DEG_neuron2, DEG_microglia2, DEG_opc2, DEG_oligodendrocyte2, DEG_endothelial2)

genes_up <- DEG_upregulated2$Gene
genes_down <- DEG_downregulated2$Gene

datasets <- list(DEG_astro2, DEG_neuron2, DEG_microglia2, DEG_opc2, DEG_oligodendrocyte2, DEG_endothelial2)

top_gene_names2 <- unlist(lapply(datasets, function(df) {
  df_sorted <- df[order(df$avg_log2FC, decreasing = TRUE), ]
  top2_up_genes <- head(df_sorted$Gene, 2)
  top2_down_genes <- tail(df_sorted$Gene, 2)
  c(top2_up_genes, top2_down_genes)
}))

combined_results2 <- combined_results2 %>%
  filter(Gene %in% genes_down | Gene %in% genes_up | Gene %in% top_gene_names2)

heatmap_data2 <- combined_results2 %>%
  select(Gene, celltype, avg_log2FC) %>%
  pivot_wider(names_from = celltype, values_from = avg_log2FC) %>%
  column_to_rownames("Gene") %>%
  as.matrix()

heatmap_data2[is.na(heatmap_data2)] <- 0

order <- c("Endothelial","Microglia", "OPCs", "Astrocytes", "Neurons", "Oligodendrocytes")

heatmap_data2 <- heatmap_data2 [, order ]

library(ComplexHeatmap)
heatmap2 <- Heatmap(
  heatmap_data2,
  col = circlize::colorRamp2(c(-7, 0, 7), c("blue", "white", "red")),
  row_names_gp = gpar(fontsize = 8),
  column_names_gp = gpar(fontsize = 8),
  name = "log2FC", row_title = "Genes", row_title_side = "right", column_title = "Cell types", column_title_side = "bottom", use_raster = T, raster_quality = 5, raster_device = "png"
)

heatmap2


```
