---
title: "Other analysis"
output: html_document
date: "2025-04-03"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(readxl)
library(ggplot2)
library(ggvenn)
library(dplyr)
```

## Once I have the DEG list, I can make different analysis to correlate the results with proteomic data.

*ACUTE CONDITION (7 dpi)*

```{r}

# VENN DIAGRAMS 
transcriptomic <- read_excel("~/R/Data/DEG_Acute.xlsx")
proteomic <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")

transcriptomic <- DEG %>% filter(abs(DEG$avg_log2FC) > 0.75)
transcriptomic <- transcriptomic$Gene

proteomic <- proteins.ME %>% filter(expr_Adjpval_Acute_Ctrl == "up" | expr_Adjpval_Acute_Ctrl == "down" )
proteomic <- proteomic$Gene

x = list("Differentially expressed genes" = transcriptomic,"Differential protein levels" = proteomic)

venn_diagram <- ggvenn(x, fill_color = c("salmon", "lightgreen"), show_elements = F, text_size = 5, set_name_size = 5)
venn_diagram

ggsave("~/R/Figures/Venn_Acute (7 dpi).png", venn_diagram, dpi = 1080, height = 7, width = 9)

intersection <- intersect(proteomic, transcriptomic)
paste(intersection,"is differentially expressed in both transcrit and protein levels when we compare uninjured and sub-chronic conditions")
```


```{r}

# CORRELATION 
transcriptomic <- DEG
transcriptomic <- transcriptomic [, c("Gene", "avg_log2FC","p_val_adj")]

proteomic <- proteins.ME
proteomic <- proteomic [, c("Gene", "logFC_Acute_Ctrl", "adj.P.Val_Acute_Ctrl")]

correlation_table <- merge(transcriptomic, proteomic, by = "Gene", all = T)

colnames(correlation_table)[2] <- "log2FC_Transcriptomics"
colnames(correlation_table)[3] <- "p_value_Transcriptomics"
colnames(correlation_table)[4] <- "log2FC_Proteomics"
colnames(correlation_table)[5] <- "p_value_Proteomics"

correlation_table$log2FC_Transcriptomics[is.na(correlation_table$log2FC_Transcriptomics)] <- 0
correlation_table$p_value_Transcriptomics[is.na(correlation_table$p_value_Transcriptomics)] <- 1

correlation_table$log2FC_Proteomics <- as.numeric(correlation_table$log2FC_Proteomics)
correlation_table$p_value_Proteomics <- as.numeric(correlation_table$p_value_Proteomics)

correlation_table <- correlation_table %>% 
  mutate(Tendence = case_when(
    log2FC_Transcriptomics > 0.75 & log2FC_Proteomics > 0.75 ~ "Both upregulated",
    log2FC_Transcriptomics < -0.75 & log2FC_Proteomics < -0.75 ~ "Both downregulated",
    log2FC_Transcriptomics > 0.75 & log2FC_Proteomics < -0.75 ~ "Transcriptomics ↑, Proteomics ↓",
    log2FC_Transcriptomics < -0.75 & log2FC_Proteomics > 0.75 ~ "Transcriptomics ↓, Proteomics ↑",
    TRUE ~ "Neutral" 
  ))

correlation_table <- correlation_table %>% 
  filter(p_value_Proteomics < 0.05 & p_value_Transcriptomics < 0.05)

correlation <- ggplot(correlation_table, aes(x = log2FC_Transcriptomics, y = log2FC_Proteomics, label = Gene)) +
  geom_point(aes(color = Tendence), size = 3, alpha = 0.7) +  
  geom_text_repel(aes(color = Tendence), size = 3) +  
  geom_smooth(method = "lm", color = "red", se = FALSE) +  
  theme_minimal() +
  labs(title = "Relationship between Gene Expression and Protein Levels",
       x = "log2FC Transcriptomics",
       y = "log2FC Proteomics") +
  scale_color_manual(values = c(
    "Both upregulated" = "green",
    "Both downregulated" = "red",
    "Transcriptomics ↑, Proteomics ↓" = "grey40",
    "Transcriptomics ↓, Proteomics ↑" = "grey20",
    "Neutral" = "grey10"
  )) +
  theme(legend.position = "bottom")
correlation
ggsave("~/R/Figures/Correlation_Acute (7 dpi).png", correlation, dpi = 1080, height = 7, width = 9)

```

*SUB-CHRONIC CONDITION (4-6 wpi)*

```{r}
# VENN DIAGRAMS 
transcriptomic <- read_excel("~/R/Data/DEG_Acute.xlsx")
proteomic <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")

transcriptomic <- DEG %>% filter(abs(DEG$avg_log2FC) > 0.75)
transcriptomic <- transcriptomic$Gene

proteomic <- proteins.ME %>% filter(expr_Adjpval_Chronic_Ctrl == "up" | expr_Adjpval_Chronic_Ctrl == "down" )
proteomic <- proteomic$Gene

x = list("Differentially expressed genes" = transcriptomic, "Differential protein levels" = proteomic)

venn_diagram <- ggvenn(x, fill_color = c("salmon", "lightgreen"), show_elements = F, text_size = 4, set_name_size = 4)
venn_diagram

ggsave("~/R/Figures/Venn_Sub-Chronic (4-6 wpi).png", venn_diagram, dpi = 1080, height = 7, width = 9)

intersection <- intersect(proteomic, transcriptomic)
paste(intersection,"is differentially expressed in both transcrit and protein levels when we compare uninjured and sub-chronic conditions")

```

```{r}
# CORRELATION
transcriptomic <- DEG
transcriptomic <- transcriptomic [, c("Gene", "avg_log2FC","p_val_adj")]

proteomic <- proteins.ME
proteomic <- proteomic [, c("Gene", "logFC_Chronic_Ctrl", "adj.P.Val_Chronic_Ctrl")]

correlation_table <- merge(transcriptomic, proteomic, by = "Gene", all = T)

colnames(correlation_table)[2] <- "log2FC_Transcriptomics"
colnames(correlation_table)[3] <- "p_value_Transcriptomics"
colnames(correlation_table)[4] <- "log2FC_Proteomics"
colnames(correlation_table)[5] <- "p_value_Proteomics"

correlation_table$log2FC_Transcriptomics[is.na(correlation_table$log2FC_Transcriptomics)] <- 0
correlation_table$p_value_Transcriptomics[is.na(correlation_table$p_value_Transcriptomics)] <- 1

correlation_table$log2FC_Proteomics <- as.numeric(correlation_table$log2FC_Proteomics)
correlation_table$p_value_Proteomics <- as.numeric(correlation_table$p_value_Proteomics)

correlation_table <- correlation_table %>% 
  mutate(Tendence = case_when(
    log2FC_Transcriptomics > 0.75 & log2FC_Proteomics > 0.75 ~ "Both upregulated",
    log2FC_Transcriptomics < -0.75 & log2FC_Proteomics < -0.75 ~ "Both downregulated",
    log2FC_Transcriptomics > 0.75 & log2FC_Proteomics < -0.75 ~ "Transcriptomics ↑, Proteomics ↓",
    log2FC_Transcriptomics < -0.75 & log2FC_Proteomics > 0.75 ~ "Transcriptomics ↓, Proteomics ↑",
    TRUE ~ "Neutral"
  ))

correlation_table <- correlation_table %>% 
  filter(p_value_Proteomics < 0.05 & p_value_Transcriptomics < 0.05)

correlation <- ggplot(correlation_table, aes(x = log2FC_Transcriptomics, y = log2FC_Proteomics, label = Gene)) +
  geom_point(aes(color = Tendence), size = 3, alpha = 0.7) +
  geom_text_repel(aes(color = Tendence), size = 3) + 
  geom_smooth(method = "lm", color = "red", se = FALSE) +  
  theme_minimal() +
  labs(title = "Relationship between Gene Expression and Protein Levels",
       x = "log2FC Transcriptomics",
       y = "log2FC Proteomics") +
  scale_color_manual(values = c(
    "Both upregulated" = "green",
    "Both downregulated" = "red",
    "Transcriptomics ↑, Proteomics ↓" = "grey40",
    "Transcriptomics ↓, Proteomics ↑" = "grey20",
    "Neutral" = "grey80"
  )) +
  theme(legend.position = "bottom")
correlation

ggsave("~/R/Figures/Correlation_Sub-Chronic (4-6 wpi).png", correlation, dpi = 1080, height = 7, width = 9)
```

