---
title: "scATAC-Seq"
output: html_document
date: "2025-04-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

```{r}
library(Seurat)
library(dplyr)
library(GenomicRanges)
library(ggplot2)
library(Matrix)
library(readxl)
library(tidyverse)
library(AnnotationDbi)
library(org.Mm.eg.db)
```

```{r}
window <- readMM("~/R/Data/ATAC-Seq/scATAC_window.mtx")
metadata_window <- read.delim("~/R/Data/ATAC-Seq/metadata_ATAC_window.txt", header = T)
features_window <- read.delim("~/R/Data/ATAC-Seq/Window_features.txt.gz", header = F)
barcodes_window <- read.delim("~/R/Data/ATAC-Seq/Window_barcodes.txt.gz", header = F)

```

```{r}
rownames(window) <- features_window$V1
colnames(window) <- barcodes_window$V1
```

```{r}
# Seurat object
seurat <- CreateSeuratObject(
  counts = window,
  assay = "ATAC",
  meta.data = metadata_window)
```

QC

```{r}
# Visualizing QC
colnames(seurat@meta.data)
VlnPlot(seurat, features = "nCount_ATAC")

# Filter
seurat <- subset(x = seurat, subset = nCount_ATAC > 9000 & nCount_ATAC < 100000)
```

Norm

```{r}
seurat <- NormalizeData(seurat)
seurat <- FindVariableFeatures(seurat)
seurat <- ScaleData(seurat)

seurat <- RunPCA(seurat, dims = 1:30)
seurat <- RunUMAP(seurat, dims = 1:30)

Idents(seurat) <- "layer1"
DimPlot(seurat, label = T)

```

```{r}
# Create gene activity matrix
Idents(seurat) <- "label"
atac <- Seurat::FindMarkers(seurat, ident.1 = "7d", ident.2 = "Uninjured", min.pct = 0.05, layer = "data")
```

```{r}
# Genes close to the regions
open_sci <- atac[atac$avg_log2FC > 1, ]
close_sci <- atac[atac$avg_log2FC < -1, ]
atac <- bind_rows(open_sci, close_sci)

# Gene name
ensembl_ids <- rownames(atac)
ensembl_ids_clean <- sub("\\..*", "", ensembl_ids)
mapped <- AnnotationDbi::select(org.Mm.eg.db,
keys = ensembl_ids_clean, keytype = "ENSEMBL", columns = c("SYMBOL"))

atac$Gene <- mapped$SYMBOL

# ECM
ecm <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")
ecm_genes <- ecm$Gene

# Filter by ECM genes
atac <- atac %>%
  dplyr::filter(Gene %in% ecm_genes)

```

Same for 2m

```{r}
# Create gene activity matrix
Idents(seurat) <- "label"
atac2 <- Seurat::FindMarkers(seurat, ident.1 = "2m", ident.2 = "Uninjured", min.pct = 0.05, layer = "data")
```

```{r}
# Genes close to the regions
open_sci2 <- rownames(atac2[atac2$avg_log2FC > 1, ])
close_sci2 <- rownames(atac2[atac2$avg_log2FC < -1, ])

# Gene name
ensembl_ids <- rownames(atac2)
ensembl_ids_clean <- sub("\\..*", "", ensembl_ids)
mapped <- AnnotationDbi::select(org.Mm.eg.db,
keys = ensembl_ids_clean, keytype = "ENSEMBL", columns = c("SYMBOL"))

mapped <- mapped[!duplicated(mapped$ENSEMBL), ]

atac2$Gene <- mapped$SYMBOL

# ECM
ecm <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")
ecm_genes <- ecm$Gene

# Filter by ECM genes
atac2 <- atac2 %>%
  dplyr::filter(Gene %in% ecm_genes)

```

OMIC INTEGRATION

```{r}
# Acute
DEG <- read_excel("~/R/Data/DEG_Acute.xlsx")
proteomic <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")
proteomic <- proteomic [, c("Gene", "logFC_Acute_Ctrl", "adj.P.Val_Acute_Ctrl")]

# Statistics
DEG <- DEG %>%
  dplyr::filter(p_val_adj < 0.05)
atac <- atac %>%
  dplyr::filter(p_val_adj < 0.05)
proteomic <- proteomic %>%
  dplyr::filter(adj.P.Val_Acute_Ctrl < 0.05)

DEG$Transcriptomics_Acute <- DEG$avg_log2FC
DEG$avg_log2FC <- NULL
atac$Accessibility_Acute <- atac$avg_log2FC
atac$avg_log2FC <- NULL
proteomic$Proteomic_Acute <- proteomic$logFC_Acute_Ctrl
proteomic$logFC_Acute_Ctrl <- NULL

DEG_ATAC <- merge(DEG, atac, by = "Gene")
DEG_ATAC <- merge(DEG_ATAC, proteomic, by = "Gene", all.x = T)

```


```{r}
# Sub-Chronic
DEG2 <- read_excel("~/R/Data/DEG_Sub-Chronic.xlsx")
proteomic <- read_excel("~/R/Data/MAIN_ECM_Proteome.xlsx")
proteomic <- proteomic [, c("Gene", "logFC_Chronic_Ctrl", "adj.P.Val_Chronic_Ctrl")]

DEG2 <- DEG2 %>%
  dplyr::filter(p_val_adj < 0.05)
atac2 <- atac2 %>%
  dplyr::filter(p_val_adj < 0.05)
proteomic <- proteomic %>%
  dplyr::filter(adj.P.Val_Chronic_Ctrl < 0.05)

DEG2$Transcriptomics_SubChronic <- DEG2$avg_log2FC
DEG2$avg_log2FC <- NULL
atac2$Accessibility_SubChronic <- atac2$avg_log2FC
atac2$avg_log2FC <- NULL
proteomic$Proteomic_SubChronic <- proteomic$logFC_Chronic_Ctrl
proteomic$logFC_Chronic_Ctrl <- NULL

DEG_ATAC2 <- merge(DEG2, atac2, by = "Gene")
DEG_ATAC2 <- merge(DEG_ATAC2, proteomic, by = "Gene", all.x = T)

```


```{r}

# Tidy 
DEG_ATAC <- DEG_ATAC %>%
  dplyr::select(matches("Gene|^Transcriptomics|^Accessibility|^Prote"))
DEG_ATAC2 <- DEG_ATAC2 %>%
  dplyr::select(matches("Gene|^Transcriptomics|^Accessibility|^Prote"))
ATAC <- merge(DEG_ATAC, DEG_ATAC2, by = "Gene", all.x = T)

ATAC <- na.omit(ATAC)

ATAC <- ATAC %>%
  mutate(across(starts_with("Transcriptomics") | starts_with("Accessibility") | starts_with("Proteomic"), as.numeric)) %>%
  pivot_longer(
    cols = starts_with("Transcriptomics") | starts_with("Accessibility") | starts_with("Proteomic"), 
    names_to = c("omic", "timePoint"), 
    names_sep = "_", 
    values_to = "log2FC"
  )

ATAC$omic <- factor(ATAC$omic, levels = c("Accessibility", "Transcriptomics", "Proteomic"))

# ATAC <- ATAC %>%
  # filter(!omic == "Proteomic")

atac_transcript <- ggplot(ATAC, aes(x = timePoint, y = log2FC, fill = omic)) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_wrap(~ Gene, scales = "free_y") + 
  theme_minimal() + 
  labs(x = "Time Point", y = "log2FC", title = "Bar plot of log2FC by Time Point and Omic", caption = "*Only the ECM genes that have been significant in the differential accessibility chromatin analysis of scATAC-Seq are shown") + 
  scale_fill_viridis_d()
atac_transcript

ggsave("~/R/Figures/ATAC_Transcriptomics_Proteomics_Windows.png", dpi = 1080, height = 7, width = 11)

```


